File: dialog/src/main/java/com/dialog/DialogFrameLayout.java
package com.dialog;

import android.content.Context;
import android.graphics.Rect;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;

public class DialogFrameLayout extends FrameLayout {

    interface OnTouchOutsideListener {
        void onTouchOutside();
    }

    GestureDetector gestureDetector = null;

    OnTouchOutsideListener onTouchOutsideListener;

    public void setOnTouchOutsideListener(OnTouchOutsideListener onTouchOutsideListener) {
        this.onTouchOutsideListener = onTouchOutsideListener;
    }

    public DialogFrameLayout(@NonNull Context context) {
        super(context);
        commonInit(context);
    }

    private void commonInit(@NonNull Context context) {
        gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
            @Override
            public boolean onDown(MotionEvent e) {
                return true;
            }

            @Override
            public boolean onSingleTapUp(MotionEvent e) {
                Rect rect = new Rect();
                getHitRect(rect);
                int count = getChildCount();
                for (int i = count - 1; i > -1; i--) {
                    View child = getChildAt(i);
                    Rect outRect = new Rect();
                    child.getHitRect(outRect);
                    if (outRect.contains((int) e.getX(), (int) e.getY())) {
                        return false;
                    }
                }
                if (onTouchOutsideListener != null) {
                    onTouchOutsideListener.onTouchOutside();
                }
                return true;
            }
        });
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return gestureDetector.onTouchEvent(event);
    }
}

File: dialog/src/main/java/com/dialog/DialogFrameLayout.kt
package com.dialog

import android.content.Context
import android.graphics.Rect
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.widget.FrameLayout

class DialogFrameLayout(context: Context) : FrameLayout(context) {
    private var onTouchOutsideListener: (() -> Unit)? = null
    private val outRect = Rect()

    fun setOnTouchOutsideListener(onTouchOutsideListener: (() -> Unit)?) {
        this.onTouchOutsideListener = onTouchOutsideListener
    }

    private val gestureDetector = GestureDetector(context, object : SimpleOnGestureListener() {
        override fun onDown(e: MotionEvent) = true

        override fun onSingleTapUp(e: MotionEvent): Boolean {
            for (i in childCount - 1 downTo 0) {
                getChildAt(i).getHitRect(outRect)
                if (outRect.contains(e.x.toInt(), e.y.toInt())) {
                    return false
                }
            }
            onTouchOutsideListener?.invoke()
            return true
        }
    })

    override fun onTouchEvent(event: MotionEvent) = gestureDetector.onTouchEvent(event)
}